app/controllers/discovered_hosts_controller.rb

```diff
@@ -80,6 +80,7 @@ def update
           delete_discovery_attribute_set(@host.id)
           process_success :success_redirect => host_path(@host), :redirect_xhr => request.xhr?
         else
+          @host.expire_token
```

Comment:
That's exactly what happens, the token is created once you hit 'submit' on the edit form because the previously discovered host causes this line to be true: https://github.com/theforeman/foreman/blob/develop/app/observers/host_observer.rb#L14
then when you cancel the token is not deleted.
I'm not sure why the transaction doesn't roll back the change, I'm guessing it's has something to do with these changes being on the managed host and not the discovered one.